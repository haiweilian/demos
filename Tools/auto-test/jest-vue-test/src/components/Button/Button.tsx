import { defineComponent } from 'vue';

// Types
import type { PropType, ButtonHTMLAttributes } from 'vue';

export type ButtonType =
  | 'default'
  | 'primary'
  | 'success'
  | 'warning'
  | 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm';

export default defineComponent({
  name: 'Button',

  props: {
    icon: String,
    round: Boolean,
    block: Boolean,
    loading: Boolean,
    loadingText: String,
    nativeType: String as PropType<ButtonHTMLAttributes['type']>,
    type: {
      type: String as PropType<ButtonType>,
      default: 'default',
    },
    size: {
      type: String as PropType<ButtonSize>,
      default: 'md',
    },
    loadingIcon: {
      type: String,
      default: 'icon-sync',
    },
  },

  emits: ['click'],

  setup(props, { emit, slots }) {
    const onClick = (event: MouseEvent) => {
      if (!props.loading) {
        emit('click', event);
      }
    };

    const getStyle = () => {
      const style = [
        'inline-flex',
        'justify-center',
        'items-center',
        'border',
        'rounded-md',
        'disabled:opacity-50',
        'disabled:cursor-not-allowed',
      ];

      // ButtonSize
      if (props.size === 'lg') {
        style.push(
          'min-w-20',
          'min-h-13',
          'px-4',
          'text-lg',
          'tracking-widest'
        );
      }
      if (props.size === 'md') {
        style.push(
          'min-w-16',
          'min-h-10',
          'px-2',
          'text-base',
          'tracking-wider'
        );
      }
      if (props.size === 'sm') {
        style.push('min-w-12', 'min-h-7', 'px-1', 'text-sm', 'tracking-wide');
      }

      // ButtonType
      if (props.block) {
        style.push('block', 'w-full');
      }
      if (props.round) {
        style.push('rounded-full');
      }
      if (props.type === 'default') {
        style.push(
          'text-dark',
          'bg-white',
          'border-gray-300',
          'active:bg-gray-200'
        );
      }
      if (props.type === 'primary') {
        style.push(
          'text-light',
          'bg-blue-500',
          'border-transparent',
          'active:bg-blue-600'
        );
      }
      if (props.type === 'success') {
        style.push(
          'text-light',
          'bg-green-500',
          'border-transparent',
          'active:bg-green-600'
        );
      }
      if (props.type === 'warning') {
        style.push(
          'text-light',
          'bg-yellow-500',
          'border-transparent',
          'active:bg-yellow-600'
        );
      }
      if (props.type === 'danger') {
        style.push(
          'text-light',
          'bg-red-500',
          'border-transparent',
          'active:bg-red-600'
        );
      }

      return style;
    };

    const renderIcon = () => {
      if (props.loading) {
        return slots.loadingIcon ? (
          slots.loadingIcon()
        ) : (
          <i
            class={[
              'iconfont',
              'inline-block',
              'animate-spin',
              props.loadingIcon,
            ]}
          />
        );
      }

      if (props.icon) {
        return <i class={['iconfont', 'inline-block', props.icon]} />;
      }

      if (slots.icon) {
        return slots.icon();
      }
    };

    const renderText = () => {
      if (props.loading) {
        if (slots.loadingText) {
          return slots.loadingText();
        }

        if (props.loadingText) {
          return <span class="mx-0.5">{props.loadingText}</span>;
        }
      }

      if (slots.default) {
        return slots.default();
      }
    };

    return () => (
      <button class={getStyle()} onClick={onClick}>
        {renderIcon()}
        {renderText()}
      </button>
    );
  },
});
