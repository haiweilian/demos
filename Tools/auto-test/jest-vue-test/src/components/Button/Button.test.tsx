import { mount } from '@vue/test-utils'
import Button from './Button'

it('测试 emit 事件', async () => {
  const wrapper = mount(Button)

  await wrapper.trigger('click');
  expect(wrapper.emitted('click')).toHaveLength(1)
})

it('测试 当 loading 不能 emit', async () => {
  const wrapper = mount(Button, {
    props: {
      loading: true
    }
  })

  await wrapper.trigger('click')
  expect(wrapper.emitted('click')).toBeUndefined()
})

it('测试渲染 icon', async () => {
  const wrapper = mount(Button, {
    props: {
      icon: 'icon-rocket',
    },
  });

  expect(wrapper.find('.icon-sync').exists()).toBeFalsy()
  expect(wrapper.find('.icon-rocket').exists()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot();

  await wrapper.setProps({
    loading: true
  })

  expect(wrapper.find('.icon-sync').exists()).toBeTruthy()
  expect(wrapper.find('.icon-rocket').exists()).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot();
})
