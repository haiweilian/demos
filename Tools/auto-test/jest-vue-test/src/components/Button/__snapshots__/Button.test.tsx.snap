// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试渲染 icon 1`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont inline-block icon-rocket\\"></i>
  <!---->
</button>"
`;

exports[`测试渲染 icon 2`] = `
"<button class=\\"inline-flex justify-center items-center border rounded-md disabled:opacity-50 disabled:cursor-not-allowed min-w-16 min-h-10 px-2 text-base tracking-wider text-dark bg-white border-gray-300 active:bg-gray-200\\"><i class=\\"iconfont inline-block animate-spin icon-sync\\"></i>
  <!---->
</button>"
`;
